### PR TITLE
Topic/imx wip

### PR DESCRIPTION
**Pull request by paulstelian97**
Local working branch, as pull request. The commits shall be sent upstream in separate branch(es)

### CODE REVIEW

File: src/drivers/imx/edma.c [hunk 1]
```diff
@@ -4,109 +4,858 @@
...
+#include <sof/lib/alloc.h>
 #include <sof/lib/dma.h>
 #include <sof/lib/io.h>
+#include <sof/platform.h>
 #include <stdint.h>
 
-static inline void edma_write(struct dma *dma, uint32_t reg, uint32_t value)
+static inline void
+edma_chan_write(struct dma_chan_data *channel, uint32_t reg, uint32_t value)
 {
```
Comment (dbaluta) [inline src/drivers/imx/edma.c]: Create generic version for dma read/write functions.

File: src/drivers/imx/edma.c [hunk 2]
```diff
@@ -4,109 +4,858 @@
...
+					 uint32_t value)
 {
-	io_reg_update_bits(dma_base(dma) + reg, mask, value);
+	io_reg_update_bits(dma_chan_base(channel->dma, channel->index) + reg,
+			   mask, value);
+}
+
+/* Hardware TCD structure. Alignment to 32 bytes is mandated by the hardware */
+struct edma_tcd {
+	uint32_t SADDR;
```
Comment (dbaluta) [inline src/drivers/imx/edma.c]: edma_hw_tcd + use lower case for members.

File: src/drivers/imx/edma.c [hunk 3]
```diff
@@ -4,109 +4,858 @@
...
+	WRITE_FIELD16(DOFF);
+	WRITE_FIELD16(CITER);
+	WRITE_FIELD(DLAST_SGA);
+	WRITE_FIELD16(BITER);
+	// Write it the last, this may trigger some action
+	WRITE_FIELD16(CSR);
+#undef WRITE_FIELD
+#undef WRITE_FIELD16
+}
+
```
Comment (dbaluta) [inline src/drivers/imx/edma.c]: drop the macros for now.

File: src/drivers/imx/edma.c [hunk 4]
```diff
@@ -4,109 +4,858 @@
...
+	READ_FIELD(DLAST_SGA);
+	READ_FIELD16(CSR);
+	READ_FIELD16(BITER);
+#undef READ_FIELD
+#undef READ_FIELD16
+}
+
+static int compute_attr(int src_width, int dest_width)
+{
+	int result = 0;
```
Comment (dbaluta) [inline src/drivers/imx/edma.c]: Use edma_ prefix for functions.

File: src/drivers/imx/edma.c [hunk 5]
```diff
@@ -4,109 +4,858 @@
...
+	}
+
+	/* Initialize channel structures */
+	channel->dma = dma;
+	channel->index = req_chan;
+
+	ch = dma_chan_get_data(channel);
+	if (!ch) {
+		// Attempt allocating the channel data
+		ch = rzalloc(RZONE_RUNTIME | RZONE_FLAG_UNCACHED,
```
Comment (dbaluta) [inline src/drivers/imx/edma.c]: Use C style comments `/* */`

File: src/drivers/imx/edma.c [hunk 6]
```diff
@@ -4,109 +4,858 @@
...
+	if (dbase + size * signum(DOFF) != sgelems->elems[1].dest)
+		return -EINVAL; /**< Not contiguous */
+	if (size != sgelems->elems[1].size)
+		return -EINVAL; /**< Mismatched sizes */
+
+	return 0; /* Ok, we good */
+}
+
+/* Some set_config helper functions */
+static int setup_tcd(struct dma_chan_data *channel, uint16_t SOFF,
```
Comment (dbaluta) [inline src/drivers/imx/edma.c]: add a proper comment to this function. Avoid uppercase names.

Comment (paulstelian97) [reply src/drivers/imx/edma.c]: All addressed.

**Comment by paulstelian97** [comment]
Closed as this branch is mostly abandoned. Will delete branch when I'm certain there is no code that can be picked up from here.

**Pull request closed without merging**